File: src/sentor_guard/sentor_guard/guard.py
"""Python implementation of SentorGuard for safe autonomous operations."""

import rclpy
from rclpy.node import Node
from rclpy.time import Time, Duration
from std_msgs.msg import String, Bool
from threading import Event, Lock
import time


class AutonomyGuardException(Exception):
    """Raised when autonomy guard conditions are not met within timeout."""
    pass


class SentorGuard:
    """
    Guard that checks robot state and autonomous mode before allowing execution.
    
    Monitors /robot_state and /autonomous_mode topics from RobotStateMachine.
    The guard ensures these messages are recent (within update_timeout).
    
    Can be used as a context manager or called directly.
    
    Example:
        # As context manager
        with SentorGuard(node):
            execute_autonomous_action()
        
        # With explicit timeout
        guard = SentorGuard(node)
        if guard.wait_for_autonomy(timeout=5.0):
            execute_autonomous_action()
    """
    
    def __init__(self, node: Node, 
                 state_topic: str = '/robot_state',
                 mode_topic: str = '/autonomous_mode',
                 update_timeout: float = 1.0,
                 required_state: str = 'active',
                 require_autonomous_mode: bool = True):
        """
        Initialize the guard.
        
        Args:
            node: ROS2 node to use for subscriptions
            state_topic: Topic publishing robot state (String: 'start-up', 'disabled', 'enabled', 'active')
            mode_topic: Topic publishing autonomous mode flag (Bool)
            update_timeout: Maximum age of state/mode messages in seconds
            required_state: State required for autonomy (default: 'active')
            require_autonomous_mode: Whether autonomous mode must be true (default: True)
        """
        self.node = node
        self.update_timeout = Duration(seconds=update_timeout)
        self.required_state = required_state
        self.require_autonomous_mode = require_autonomous_mode
        
        self._lock = Lock()
        self._current_state = None
        self._autonomous_mode = None
        self._last_state_time = None
        self._last_mode_time = None
        self._condition_met = Event()
        
        # Subscribe to state and mode topics
        self._state_sub = node.create_subscription(
            String,
            state_topic,
            self._state_callback,
            10
        )
        
        self._mode_sub = node.create_subscription(
            Bool,
            mode_topic,
            self._mode_callback,
            10
        )
        
        self.node.get_logger().info(
            f"SentorGuard initialized: required_state='{required_state}', "
            f"update_timeout={update_timeout}s"
        )
    
    def _state_callback(self, msg):
        """Handle robot state updates."""
        with self._lock:
            self._current_state = msg.data
            self._last_state_time = self.node.get_clock().now()
            self._check_conditions()
    
    def _mode_callback(self, msg):
        """Handle autonomous mode updates."""
        with self._lock:
            self._autonomous_mode = msg.data
            self._last_mode_time = self.node.get_clock().now()
            self._check_conditions()
    
    def _check_conditions(self):
        """Check if all conditions are met."""
        now = self.node.get_clock().now()
        
        # Check if we have received state message
        if self._current_state is None or self._last_state_time is None:
            self._condition_met.clear()
            return
        
        # Check if state message is recent
        state_age = now - self._last_state_time
        if state_age > self.update_timeout:
            self._condition_met.clear()
            return
        
        # Check state value
        if self._current_state != self.required_state:
            self._condition_met.clear()
            return
        
        # Check if we have received mode message
        if self._autonomous_mode is None or self._last_mode_time is None:
            self._condition_met.clear()
            return
        
        # Check if mode message is recent
        mode_age = now - self._last_mode_time
        if mode_age > self.update_timeout:
            self._condition_met.clear()
            return
        
        # Check autonomous mode
        if self.require_autonomous_mode and not self._autonomous_mode:
            self._condition_met.clear()
            return
        
        # All conditions met
        self._condition_met.set()
    
    def is_autonomy_allowed(self) -> bool:
        """
        Check if autonomy is currently allowed (non-blocking).
        
        Returns:
            True if all guard conditions are satisfied, False otherwise
        """
        with self._lock:
            self._check_conditions()  # Recheck heartbeat age
            return self._condition_met.is_set()
    
    def get_blocking_reason(self) -> str:
        """
        Get human-readable reason why autonomy is blocked.
        
        Returns:
            String describing why autonomy is not allowed
        """
        with self._lock:
            now = self.node.get_clock().now()
            
            # Check state
            if self._current_state is None or self._last_state_time is None:
                return "Robot state not received"
            
            state_age = now - self._last_state_time
            if state_age > self.update_timeout:
                return f"Robot state stale ({state_age.nanoseconds / 1e9:.2f}s old)"
            
            if self._current_state != self.required_state:
                return f"State is '{self._current_state}', required '{self.required_state}'"
            
            # Check mode
            if self._autonomous_mode is None or self._last_mode_time is None:
                return "Autonomous mode not received"
            
            mode_age = now - self._last_mode_time
            if mode_age > self.update_timeout:
                return f"Autonomous mode stale ({mode_age.nanoseconds / 1e9:.2f}s old)"
            
            if self.require_autonomous_mode and not self._autonomous_mode:
                return "Autonomous mode is disabled"
            
            return "Unknown reason"
    
    def wait_for_autonomy(self, timeout: float = None) -> bool:
        """
        Wait until autonomy is allowed.
        
        Args:
            timeout: Maximum time to wait in seconds. None for indefinite.
            
        Returns:
            True if autonomy is allowed, False if timeout occurred
        """
        start_time = time.time()
        
        while rclpy.ok():
            if self.is_autonomy_allowed():
                return True
            
            if timeout is not None:
                elapsed = time.time() - start_time
                if elapsed >= timeout:
                    reason = self.get_blocking_reason()
                    self.node.get_logger().warn(
                        f"Autonomy not granted within {timeout}s: {reason}"
                    )
                    return False
            
            # Spin once to process callbacks
            rclpy.spin_once(self.node, timeout_sec=0.1)
            
        return False
    
    def __enter__(self):
        """Context manager entry - waits indefinitely by default."""
        if not self.wait_for_autonomy():
            raise AutonomyGuardException("Autonomy not allowed and node is shutting down")
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        return False
    
    def guarded_wait(self, timeout: float = None):
        """
        Wait for autonomy with timeout, raising exception on failure.
        
        Args:
            timeout: Maximum time to wait. None for indefinite.
            
        Raises:
            AutonomyGuardException: If timeout occurs
        """
        if not self.wait_for_autonomy(timeout):
            reason = self.get_blocking_reason()
            if timeout:
                raise AutonomyGuardException(
                    f"Autonomy not granted within {timeout}s timeout: {reason}"
                )
            else:
                raise AutonomyGuardException(
                    f"Autonomy not allowed: {reason}"
                )


def sentor_guarded(guard: SentorGuard = None, timeout: float = None):
    """
    Decorator that ensures a function only executes when guard conditions are met.
    
    This decorator can be used in two ways:
    1. With a guard instance as an argument
    2. As a method decorator in a class that has a 'guard' attribute
    
    Args:
        guard: SentorGuard instance to check (optional if used on class methods)
        timeout: Maximum time to wait for guard conditions (None = indefinite)
    
    Example:
        # Using with explicit guard
        @sentor_guarded(guard=my_guard, timeout=5.0)
        def autonomous_action():
            execute_navigation()
        
        # Using as method decorator (class must have self.guard)
        class MyNode(Node):
            def __init__(self):
                self.guard = SentorGuard(self)
            
            @sentor_guarded()
            def autonomous_action(self):
                execute_navigation()
    
    Raises:
        AutonomyGuardException: If guard conditions are not met within timeout
    """
    def decorator(func):
        def wrapper(*args, **kwargs):
            # Try to get guard from arguments
            guard_instance = guard
            
            # If no guard provided, try to get it from self (for class methods)
            if guard_instance is None and len(args) > 0:
                # Check if first argument (self) has a guard attribute
                if hasattr(args[0], 'guard'):
                    guard_instance = args[0].guard
            
            # If still no guard, raise an error
            if guard_instance is None:
                raise ValueError(
                    "sentor_guarded decorator requires either a 'guard' parameter "
                    "or to be used on a method of a class with a 'guard' attribute"
                )
            
            # Check if guard conditions are met
            if not isinstance(guard_instance, SentorGuard):
                raise TypeError(
                    f"Expected SentorGuard instance, got {type(guard_instance)}"
                )
            
            # Wait for autonomy with the specified timeout
            guard_instance.guarded_wait(timeout=timeout)
            
            # Execute the function if guard conditions are met
            return func(*args, **kwargs)
        
        return wrapper
    return decorator

File: src/sentor_guard/src/guard.cpp
#include "sentor_guard/guard.hpp"
#include "sentor_guard/msg/guard_status.hpp"
#include <sstream>

namespace sentor_guard {

SentorGuard::SentorGuard(rclcpp::Node::SharedPtr node)
    : SentorGuard(node, Options()) {
}

SentorGuard::SentorGuard(rclcpp::Node::SharedPtr node, const Options& options)
    : node_(node), options_(options),
      last_state_time_(node->get_clock()->now()),
      last_mode_time_(node->get_clock()->now()) {
    
    // Create subscriptions
    state_sub_ = node_->create_subscription<std_msgs::msg::String>(
        options_.state_topic, 10,
        std::bind(&SentorGuard::stateCallback, this, std::placeholders::_1));
    
    mode_sub_ = node_->create_subscription<std_msgs::msg::Bool>(
        options_.mode_topic, 10,
        std::bind(&SentorGuard::modeCallback, this, std::placeholders::_1));
    
    // Create publisher for blocking status
    status_publisher_ = node_->create_publisher<sentor_guard::msg::GuardStatus>(
        "/sentor_guard/blocking_reason", 10);
    
    RCLCPP_INFO(node_->get_logger(),
        "SentorGuard initialized: required_state='%s', update_timeout=%ldms",
        options_.required_state.c_str(), options_.update_timeout.count());
}

void SentorGuard::stateCallback(const std_msgs::msg::String::SharedPtr msg) {
    std::lock_guard<std::mutex> lock(mutex_);
    current_state_ = msg->data;
    last_state_time_ = node_->get_clock()->now();
    checkConditions();
}

void SentorGuard::modeCallback(const std_msgs::msg::Bool::SharedPtr msg) {
    std::lock_guard<std::mutex> lock(mutex_);
    autonomous_mode_ = msg->data;
    last_mode_time_ = node_->get_clock()->now();
    checkConditions();
}

void SentorGuard::checkConditions() {
    auto now = node_->get_clock()->now();
    
    // Check if we have received state message
    if (current_state_.empty()) {
        condition_met_ = false;
        return;
    }
    
    // Check if state message is recent
    auto state_age = now - last_state_time_;
    if (state_age > rclcpp::Duration(options_.update_timeout)) {
        condition_met_ = false;
        return;
    }
    
    // Check state value
    if (current_state_ != options_.required_state) {
        condition_met_ = false;
        return;
    }
    
    // Check if mode message is recent
    auto mode_age = now - last_mode_time_;
    if (mode_age > rclcpp::Duration(options_.update_timeout)) {
        condition_met_ = false;
        return;
    }
    
    // Check autonomous mode
    if (options_.require_autonomous_mode && !autonomous_mode_) {
        condition_met_ = false;
        return;
    }
    
    // All conditions met
    bool was_met = condition_met_;
    condition_met_ = true;
    
    if (!was_met && condition_met_) {
        cv_.notify_all();
    }
}

bool SentorGuard::isAutonomyAllowed() {
    std::lock_guard<std::mutex> lock(mutex_);
    checkConditions();  // Recheck heartbeat age
    return condition_met_;
}

std::string SentorGuard::getBlockingReason() const {
    std::lock_guard<std::mutex> lock(mutex_);
    auto now = node_->get_clock()->now();
    
    // Check state
    if (current_state_.empty()) {
        return "Robot state not received";
    }
    
    auto state_age = now - last_state_time_;
    if (state_age > rclcpp::Duration(options_.update_timeout)) {
        return "Robot state stale (" + std::to_string(state_age.seconds()) + "s old)";
    }
    
    if (current_state_ != options_.required_state) {
        return "State is '" + current_state_ + "', required '" + options_.required_state + "'";
    }
    
    // Check mode
    auto mode_age = now - last_mode_time_;
    if (mode_age > rclcpp::Duration(options_.update_timeout)) {
        return "Autonomous mode stale (" + std::to_string(mode_age.seconds()) + "s old)";
    }
    
    if (options_.require_autonomous_mode && !autonomous_mode_) {
        return "Autonomous mode is disabled";
    }
    
    return "Unknown reason";
}

std::vector<std::string> SentorGuard::getTruncatedCallStack(int max_frames) {
    std::vector<std::string> result;
    
    #ifdef __GNUC__
    void* addresses[max_frames + 5];  // Extra frames to skip
    int size = backtrace(addresses, max_frames + 5);
    char** symbols = backtrace_symbols(addresses, size);
    
    // Skip first 3 frames (this function and guard methods)
    for (int i = 3; i < std::min(size, max_frames + 3); ++i) {
        std::string frame = symbols[i];
        
        // Try to demangle C++ symbols
        size_t begin = frame.find('(');
        size_t end = frame.find('+', begin);
        if (begin != std::string::npos && end != std::string::npos) {
            std::string mangled = frame.substr(begin + 1, end - begin - 1);
            int status;
            char* demangled = abi::__cxa_demangle(mangled.c_str(), nullptr, nullptr, &status);
            if (status == 0 && demangled) {
                result.push_back(demangled);
                free(demangled);
            } else {
                result.push_back(frame);
            }
        } else {
            result.push_back(frame);
        }
    }
    
    free(symbols);
    #else
    result.push_back("Call stack not available on this platform");
    #endif
    
    return result;
}

void SentorGuard::publishBlockingStatus(bool is_blocking, const std::vector<std::string>& call_stack) {
    if (!status_publisher_) {
        return;
    }
    
    try {
        sentor_guard::msg::GuardStatus msg;
        msg.node_name = node_->get_name();
        msg.is_blocking = is_blocking;
        
        if (is_blocking) {
            msg.blocking_reason = getBlockingReason();
            msg.call_stack = call_stack;
            auto now = node_->get_clock()->now();
            msg.blocked_at = now;
            msg.blocked_duration = 0.0;
        } else {
            // Guard is passing after being blocked
            msg.blocking_reason = "";
            msg.call_stack.clear();
            if (is_currently_blocking_) {
                auto now = node_->get_clock()->now();
                auto duration = now - blocking_start_time_;
                msg.blocked_duration = duration.seconds();
                msg.blocked_at = blocking_start_time_;
            } else {
                msg.blocked_duration = 0.0;
                msg.blocked_at = rclcpp::Time(0);
            }
        }
        
        status_publisher_->publish(msg);
        
    } catch (const std::exception& e) {
        RCLCPP_WARN(node_->get_logger(), "Failed to publish blocking status: %s", e.what());
    }
}

bool SentorGuard::waitForAutonomy(std::chrono::milliseconds timeout) {
    auto start = std::chrono::steady_clock::now();
    bool first_block = true;
    
    while (rclcpp::ok()) {
        {
            std::unique_lock<std::mutex> lock(mutex_);
            checkConditions();
            
            if (condition_met_) {
                // If we were previously blocking, publish that we're now passing
                if (is_currently_blocking_) {
                    publishBlockingStatus(false);
                    is_currently_blocking_ = false;
                    blocking_call_stack_.clear();
                }
                return true;
            }
            
            // We're blocking - publish status on first block
            if (first_block) {
                first_block = false;
                is_currently_blocking_ = true;
                blocking_start_time_ = node_->get_clock()->now();
                blocking_call_stack_ = getTruncatedCallStack();
                publishBlockingStatus(true, blocking_call_stack_);
            }
            
            if (timeout.count() > 0) {
                auto elapsed = std::chrono::steady_clock::now() - start;
                auto remaining = timeout - std::chrono::duration_cast<std::chrono::milliseconds>(elapsed);
                
                if (remaining.count() <= 0) {
                    auto reason = getBlockingReason();
                    RCLCPP_WARN(node_->get_logger(),
                        "Autonomy not granted within %ldms: %s",
                        timeout.count(), reason.c_str());
                    return false;
                }
                
                cv_.wait_for(lock, std::chrono::milliseconds(100));
            } else {
                cv_.wait_for(lock, std::chrono::milliseconds(100));
            }
        }
        
        // Spin once to process callbacks
        rclcpp::spin_some(node_);
    }
    
    return false;
}

void SentorGuard::guardedWait(std::chrono::milliseconds timeout) {
    if (!waitForAutonomy(timeout)) {
        auto reason = getBlockingReason();
        if (timeout.count() > 0) {
            throw AutonomyGuardException(
                "Autonomy not granted within " + std::to_string(timeout.count()) + 
                "ms timeout: " + reason);
        } else {
            throw AutonomyGuardException("Autonomy not allowed: " + reason);
        }
    }
}

} // namespace sentor_guard
